digid_eherkenning/models/certificates.py

```diff
@@ -0,0 +1,201 @@
+"""
+Functionality to relate one or more certificates to a SAMLv2 configuration.
+"""
+
+from __future__ import annotations
+
+import logging
+from typing import TYPE_CHECKING, TypeAlias
+
+from django.db import models
+from django.utils import timezone
+from django.utils.translation import gettext_lazy as _
+
+from simple_certmanager.constants import CertificateTypes
+from simple_certmanager.models import Certificate
+
+from ..choices import ConfigTypes
+
+if TYPE_CHECKING:
+    from .digid import DigidConfiguration
+    from .eherkenning import EherkenningConfiguration
+
+logger = logging.getLogger(__name__)
+
+_AnyDigiD: TypeAlias = "type[DigidConfiguration] | DigidConfiguration"
+_AnyEH: TypeAlias = "type[EherkenningConfiguration] | EherkenningConfiguration"
+
+
+class ConfigCertificateQuerySet(models.QuerySet["ConfigCertificate"]):
+    def for_config(self, config: _AnyDigiD | _AnyEH):
+        config_type = config._as_config_type()
+        return self.filter(config_type=config_type)
+
+    def select_certificates(self) -> tuple[Certificate, Certificate | None]:
+        """
+        Select the best candidates for the current and next certificate.
+
+        The certificates are used for signing authentication requests (and metadata)
+        itself, and the possible certificates that perform signing are included in the
+        metadata. For zero-downtime/gradual replacement, a current and next certificate
+        can be provided (this is a limitation in python3-saml).
+
+        We look for the current certificate and the next with the following algorithm:
+
+        * order candidates by valid_from, so we favour existing/the oldest keypairs
+        * order candidates by expiry date, so if they have an identical valid_from, we
+          favour the one that will expiry first (the other one(s) automatically become
+          the next certificate
+        * discard any candidates that do not meet our key pair requirements, ignoring
+          valid_from/until
```

Comment:
Yeah, to be fair, I started updating our code before I knew that python3-saml only supports 2 certificates - I had wrongly assumed that it could be an arbitrary amount.

This setup should make life the easiest for admins - the old expired certificates do not necessarily have to be cleaned up, they will just not be selected, and possibly you can already prepare many certificates ahead of time if you'd want, it won't break anything.

Looking at an upload date seems fragile to me:

* the upload date does not say anything about validity of the certificate itself
* people may upload them in the wrong order because they made a mistake (selected wrong file on disk for example)
* we still need to explicitly assign certificates that may be used for DigiD, as there may be certificates in the system used for other things (like zgw-consumers), so either way we need to narrow the set of candidates already.
* there's various ways certificates can break - after configuring everyone, someone could remove the private key, change the type of the certificate... which we cannot prevent with database checks/constraints. this approach should be the most robust in messy environments

Reply:
I see!